core/unit_test/incremental/Test10_HierarchicalBasics.hpp

```diff
@@ -0,0 +1,90 @@
+//              Copyright (2014) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+// @Kokkos_Feature_Level_Required:1
+// Unit test for hierarchical parallelism
+// Create concurrent work hierarchically and verify if
+// contributions of paticipating processing units corresponds to expected value
+
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+
+// Degrees of concurrency per nesting level
+#define N 16
+#define M 16
+
+namespace Test {
+
+typedef int value_type;
+
+template <class ExecSpace>
+struct Functor {
+  typedef Kokkos::TeamPolicy<ExecSpace> TeamPolicy;
+  typedef typename TeamPolicy::member_type member_type;
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const member_type& team, value_type& update) const {
+    int m = team.league_rank() + team.team_rank();
+    team.team_reduce(Kokkos::Max<int>(m));
+    if (m != team.league_rank() + (team.team_size() - 1)) update++;
+  }
+};
+
+template <class ExecSpace>
+struct HierarchicalBasics {
+  typedef Kokkos::TeamPolicy<ExecSpace> TeamPolicy;
+  typedef Test::Functor<ExecSpace> Functor;
+
+  void run() {
+    int error = 0;
+    Kokkos::parallel_reduce("Teams", TeamPolicy(N, M), Functor(), error);
```

Comment:
According to https://github.com/kokkos/internal-documents/blob/master/developer-notes/Backend-Development-Guide.md this should only happen in Test 15. Test 10 should only to parallel for and NO nested things or reductions.